### PR TITLE
[Diffusion][SenseNova] Build RoPE tables once per forward, not per layer

### PR DESCRIPTION
**Pull request by wes-lyu**
## Motivation

`Qwen3Attention` computes the t/h/w RoPE cos/sin tables in every layer, although the positions and RoPE config are the same across layers. For the 42-layer dense backbone, this change reduces table construction from 126 calls to 3 per forward.

## Modifications

Build the tables once in `Qwen3Model.forward` using the first layer's RoPE modules, then pass them through the decoder layers to attention. Attention falls back to computing its own tables when none are supplied, preserving the existing MoE and direct-call paths.

## Accuracy Tests

- `python -m pytest python/sglang/multimodal_gen/test/unit/test_sensenova_u1.py -q`: 57 passed. Tests cover matching tables across layers, construction counts on generation and understanding paths, and bit-identical model output with sharing enabled and bypassed.
- Fixed-seed image generation produced byte-identical PNGs before and after the change at all sizes and step counts below.

Generation command used for the comparison:

```bash
for size in 256 512 1024 2048; do
  for steps in 4 50; do
    sglang generate --model-path /path/to/SenseNova-U1.5-8B-MoT --prompt "a red cube on a table" \
      --height $size --width $size --num-inference-steps $steps --seed 42 --num-gpus 1 --save-output
  done
done
```

Runs alternated between the parent commit and this change; output PNG SHA-256 hashes were compared.

## Speed Tests and Profiling

RTX 4090 (48 GB), Xeon Gold 6530 (8 physical cores). Before/after runs were interleaved within one session, with 3 rounds per configuration; the table reports median times.

| Size | 4 steps, before → after | 50 steps, before → after | Latency reduction (50 steps) |
|---|---|---|---|
| 256×256 | 2.20 → 1.86 s | 11.01 → 7.34 s | 33% |
| 512×512 | 2.26 → 1.86 s | 11.00 → 7.34 s | 33% |
| 1024×1024 | 2.60 → 2.50 s | 14.90 → 14.68 s | 1.5% |
| 2048×2048 | 6.76 → 6.68 s | 61.58 → 61.47 s | 0.2% |

The improvement is concentrated at 256×256 and 512×512 on this setup. GPU utilization during denoising was 70–79% at 512×512 and 95–100% at 1024×1024, consistent with reduced CPU overhead helping more at smaller sizes.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #34830163669](https://github.com/sgl-project/sglang/actions/runs/34830163669)
Latest PR Test (Extra): :x: [Run #34830163445](https://github.com/sgl-project/sglang/actions/runs/34830163445)
Latest PR Test (AMD ROCm 10): :x: [Run #34830163604](https://github.com/sgl-project/sglang/actions/runs/34830163604)

### CODE REVIEW

**Comment by wes-lyu** [comment]
Verified on a 4090 (48 GB), single GPU, seed 42, against the parent commit. The saved images are byte-identical at every configuration (2048x2048/50 `74022f66000f…`, 512x512/20 `b7a2bc2b2b04…`, 256x256/50 one hash across all eight runs, `e68b6b6d336b…`). The added tests cover the same equivalence below the image: the tables each layer would build are asserted bitwise equal, and the model output is asserted exactly equal between the shared path and the path where every layer rebuilds its own. Those hashes reproduce on this machine; on a different GPU or library version I would expect numerical equivalence rather than bit-identical floats.

The speedup depends on resolution. 256x256/50 steps, three alternating runs each: base 11.54 / 11.01 / 11.02 s against 7.19 / 7.18 / 7.27 s, about 1.55x. 512x512/20 steps, two runs each: 5.26 / 5.28 s against 3.72 / 3.79 s, about 1.40x. 2048x2048/50 steps does not move: 61.70 vs 61.66 s.

That fits the mechanism. A decoder layer issues roughly 112 kernels and ~33 of them are the three RoPE table builds, so removing the rebuilds takes about a third of the launches out of each layer. At small resolutions the step is launch-bound and that cut lands on the critical path; at 2048x2048 the GPU is saturated and the hidden CPU work changes nothing. That is also the model's default resolution.

File: python/sglang/multimodal_gen/test/unit/test_sensenova_u1.py [hunk 1]
```diff
@@ -849,3 +849,153 @@ def test_sensenova_u1_multi_output_entrypoint_mixed_failure_fails_parent(
...
+    from sglang.multimodal_gen.runtime.models.sensenova_u1.neo_unify.modeling_qwen3 import (
+        Qwen3Attention,
+        Qwen3Model,
+    )
+
+    torch.manual_seed(0)
+    model = Qwen3Model(_tiny_dense_config()).eval()
+    embeds = torch.randn(1, 5, _HIDDEN_DIM)
+    indicators = torch.ones(1, 5, dtype=torch.bool)
+    indexes = torch.zeros(3, 5, dtype=torch.long)
```
Comment (rwang5203) [inline python/sglang/multimodal_gen/test/unit/test_sensenova_u1.py:978]: Zero positions make RoPE an identity, so this parity test misses incorrect axis forwarding. Consider using distinct nonzero temporal, height, and width positions.

File: python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py [hunk 2]
```diff
@@ -1388,7 +1415,17 @@ def forward(
...
 
-        for decoder_layer in self.layers[: self.config.num_hidden_layers]:
+        # The tables depend only on the positions, plus the activation dtype and
+        # device that `Qwen3RotaryEmbedding.forward` reads off `x`. All three are
+        # fixed for the whole forward, so one build serves every layer.
+        layers = self.layers[: self.config.num_hidden_layers]
+        position_embeddings = None
+        if layers:
+            position_embeddings = layers[0].self_attn._resolve_rope_tables(
+                indexes, hidden_states
```
Comment (rwang5203) [inline python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py:1432]: With fp32 weights and bf16 autocast, denoising builds bf16 tables before normalization promotes activations, changing outputs. Consider rebuilding tables when their dtype differs from the normalized activations.

File: python/sglang/multimodal_gen/test/unit/test_sensenova_u1.py (hunk 1 above)
Comment (wes-lyu) [reply python/sglang/multimodal_gen/test/unit/test_sensenova_u1.py]: Thanks, updated the parity test to use different position steps for each axis, so height/width swaps are no longer hidden by all-zero positions.

File: python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py (hunk 2 above)
Comment (wes-lyu) [reply python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py:1432]: Thanks, added a dtype check to rebuild the tables when their dtype differs from the normalized activations. Added a regression test with FP32 weights and BF16 autocast.
